### PR TITLE
Add the ability to load niftis

### PR DESCRIPTION
**Pull request by gab-berest**
Integration of the load_nifti function that uses the read_nii functionality to load Nifti type data folders. Linked ot issue #13.

### CODE REVIEW

File: .idea/misc.xml
Comment (jcohenadad) [inline .idea/misc.xml:1]: all these files shouldn't go into the repository

File: shimmingtoolbox/cli/load_nifti.py
Comment (jcohenadad) [inline shimmingtoolbox/cli/load_nifti.py:1]: not sure what is the purpose of this CLI?

File: shimmingtoolbox/dicom_to_nifti.py
Comment (jcohenadad) [inline shimmingtoolbox/dicom_to_nifti.py:1]: that's a duplicate of https://github.com/shimming-toolbox/shimming-toolbox-py/pull/18

having several unmerged branches that share the same development makes it difficult to keep track and to review (we end up making the same comments on both PRs...). is this PR necessary? or can it be re-focused?

Comment (gab-berest) [reply shimmingtoolbox/dicom_to_nifti.py:1]: It's just that I'm using the content in this functionality. I most say, I would probably put it with the other PR since it is closely related. It was a separate task so I made a separate PR, but I think it would be better to put it together with the other one.

File: .idea/misc.xml
Comment (gab-berest) [reply .idea/misc.xml:1]: These were pulled from #18 . I'll check to remove them and add them to .gitignore on the pulled content.

File: shimmingtoolbox/load_nifti.py
Comment (po09i) [inline shimmingtoolbox/load_nifti.py:16]: Variable convention : https://github.com/neuropoly/spinalcordtoolbox/wiki/variables#variables
`path_acquisition` or `path_data`

File: shimmingtoolbox/read_nii.py
Comment (jcohenadad) [inline shimmingtoolbox/read_nii.py:8]: is this function needed? `nibabel.load()` outputs the image + header (which contains the metadata)

File: shimmingtoolbox/read_nii.py
Comment (jcohenadad) [inline shimmingtoolbox/read_nii.py:40]: this should be already solved by the BIDS convention in #18

File: test/test_load_nifti.py
Comment (jcohenadad) [inline test/test_load_nifti.py:1]: i would suggest to wait for the dev core team to agree on the proposed developments before implementing the unit tests (otherwise, the unit tests might become stale/irrelevant)

File: shimmingtoolbox/read_nii.py
Comment (gab-berest) [reply shimmingtoolbox/read_nii.py:8]: I am going to remove it. It is just convenient since it was already programmed with the verifications.

File: test/test_load_nifti.py
Comment (gab-berest) [reply test/test_load_nifti.py:1]: Sure, I implemented them for now so I can validate that the core functionality of the function works correctly, but I will wait for further information.

File: shimmingtoolbox/load_nifti.py
Comment (jcohenadad) [inline shimmingtoolbox/load_nifti.py:94]: i think pep8 asks for no space between equality terms in input arguments

File: shimmingtoolbox/load_nifti.py
Comment (po09i) [inline shimmingtoolbox/load_nifti.py:66]: This correctly implements the matlab function but it should be revisited as the number of files is not always the number of echoes. What if there are multiple runs, multiple time points, etc. 

The header should probably be read to figure out the appropriate way to "stitch" the files together.

File: shimmingtoolbox/load_nifti.py
Comment (po09i) [inline shimmingtoolbox/load_nifti.py:128]: Can nibabel output this for us instead?

File: shimmingtoolbox/load_nifti.py
Comment (jidicula) [inline shimmingtoolbox/load_nifti.py:48]: Copied from: https://github.com/shimming-toolbox/shimming-toolbox-py/pull/46#discussion_r465966701

Since our `setup.py` explicitly requires Python 3.6 or greater, we have access to f-strings for string formatting, which is a bit tidier for future readers.

See these resources for details:
- https://realpython.com/python-f-strings/#f-strings-a-new-and-improved-way-to-format-strings-in-python
- https://docs.python.org/3.6/reference/lexical_analysis.html#f-strings

File: shimmingtoolbox/load_nifti.py
Comment (jidicula) [inline shimmingtoolbox/load_nifti.py:61]: See my comment above about f-strings

File: shimmingtoolbox/load_nifti.py
Comment (jidicula) [inline shimmingtoolbox/load_nifti.py:69]: f-string

**Review by jidicula — CHANGES_REQUESTED** [review]
This is a quick review to catch instances where `str.format()` could be refactored into the cleaner f-string method of string formatting.

File: shimmingtoolbox/load_nifti.py
Comment (gab-berest) [reply shimmingtoolbox/load_nifti.py:66]: True. I should add the ability to detect the correct way to parse through data. From what I've seen for multiple runs, the files are in separate folders which you can choose before hand. But, what do you mean by time points?

File: shimmingtoolbox/load_nifti.py
Comment (gab-berest) [reply shimmingtoolbox/load_nifti.py:128]: I didn't see a image type field in the nibabel header and nibabel cannot open json type files. So the information has to be loaded with another way than nibabel for what I've seen. Maybe I've just missed the field to do it.

File: shimmingtoolbox/load_nifti.py
Comment (gab-berest) [reply shimmingtoolbox/load_nifti.py:66]: If you mean that we should parse the description of the header file, is it really reliable as information?

File: shimmingtoolbox/load_nifti.py
Comment (jcohenadad) [reply shimmingtoolbox/load_nifti.py:128]: @po09i what do we want from this `image_type()` function? what would be example of output "type"?

Comment (gab-berest) [reply shimmingtoolbox/load_nifti.py:128]: It only tells us if it's phase or magnitude from the 'ImageType' section in the json file thta is linked to the nifti file. The read_nii function uses it to know if the data needs to be scaled in radian or not. I did not find a good way in the header of the nibabel file to tell us this information.

Comment (jcohenadad) [reply shimmingtoolbox/load_nifti.py:128]: don't we already have that information after [BIDS conversion](https://github.com/shimming-toolbox/shimming-toolbox-py/blob/master/config/dcm2bids.json#L105)?

Comment (gab-berest) [reply shimmingtoolbox/load_nifti.py:128]: Yes, but dcm2bids automatically creates the json file that contain this information with the data. So I was trying to make this function independent form dcm2bids so when you only have the original nifti files and json without dcm2bids, you could also know this information.

Comment (jcohenadad) [reply shimmingtoolbox/load_nifti.py:128]: ah i see. I think at this point in the development i would suggest to not bother covering all possible scenario. It is fair to assume input data is either DICOM or BIDS folder, in which case no need for the `image_type()` function.

File: shimmingtoolbox/load_nifti.py
Comment (gab-berest) [reply shimmingtoolbox/load_nifti.py:66]: @po09i  I think that the NIFTI files generated by dcm2bids will be always as different echoes for each file and the files will contain the rest of the data. Tell me if I'm wrong. I have not seen another example on the BIDS website.

Comment (po09i) [reply shimmingtoolbox/load_nifti.py:66]: I found this link to the [bids convention](https://bids-specification.readthedocs.io/en/stable/04-modality-specific-files/01-magnetic-resonance-imaging-data.html#the-run-entity) really useful. It describes what runs are and other specific bids format information. 

We used to separate scans in separate acquisitions however now that we are trying to deal with only having one config file, it is not a guaranty that one acquisition will be in the folder (If we run 2 fieldmaps for example). I think counting the number of files in a folder is not a rigorous check and the only way to make sure is indeed reading the header info.

Comment (jcohenadad) [reply shimmingtoolbox/load_nifti.py:66]: let's not spend too much time on this, as it is still unclear how the typical experimental workflow will look like (e.g. will we impose convention for naming runs during acquisition? will we only select the lastest fieldmap? etc.). For now, let's keep it simple and as experimental scenario are clearer (meaning: post-covid19: we can scan again) we will revisit this

**Review by gaspardcereza — COMMENTED** [review]
Everything seems to be working with the B0 niftis. 
It doesn't work with my B1 maps because the dimensions of my NIfTI aren't the same  (their dimension are [x, y, slice, n_coils]). I'll raise the issue once this PR is merged but I think the easiest solution will be to have a separate fonction dedicated to B1 niftis

File: shimmingtoolbox/load_nifti.py
Comment (gab-berest) [reply shimmingtoolbox/load_nifti.py:48]: Oups. Forgot to remove it.

**Comment by gab-berest** [comment]
> 
> 
> Everything seems to be working with the B0 niftis.
> It doesn't work with my B1 maps because the dimensions of my NIfTI aren't the same (their dimension are [x, y, slice, n_coils]). I'll raise the issue once this PR is merged but I think the easiest solution will be to have a separate fonction dedicated to B1 niftis

What are your data files so I can test. Normally, it would, with these dimensions, create a 5D array of [x,y,slices(as z),echoes,n_choil]. Maybe an argument could be given to load B0 or B1 maps or just by header information.

**Comment by gaspardcereza** [comment]
> > Everything seems to be working with the B0 niftis.
> > It doesn't work with my B1 maps because the dimensions of my NIfTI aren't the same (their dimension are [x, y, slice, n_coils]). I'll raise the issue once this PR is merged but I think the easiest solution will be to have a separate fonction dedicated to B1 niftis
> 
> What are your data files so I can test. Normally, it would, with these dimensions, create a 5D array of [x,y,slices(as z),echoes,n_choil]. Maybe an argument could be given to load B0 or B1 maps or just by header information.

I used the data that are in the `b1_maps` folder in the testing-data repository. But I convert them using the `dicom_to_nifti` function for B1 maps that is not merged yet (still some work to do on them).
Even when I try `load_nifti` after a simple call to dcm2niix on my `b1_maps` folder as follows: 
```
loaded_nifti = load_nifti.load_nifti(r'C:\Users\gaspa\Documents\GitHub\shimming-toolbox-py\testing_data\b1_maps')
``` 
It returns a 5d array that is a 64x2x8x2x64 array instead of (64x64x2x16 in the original nifti). 
```
loaded_nifti[0].shape
(64, 2, 16, 1, 64)
```

**Comment by gab-berest** [comment]
> 
> 
> > What are your data files so I can test. Normally, it would, with these dimensions, create a 5D array of [x,y,slices(as z),echoes,n_choil]. Maybe an argument could be given to load B0 or B1 maps or just by header information.
> 
> I used the data that are in the `b1_maps` folder in the testing-data repository. But I convert them using the `dicom_to_nifti` function for B1 maps that is not merged yet (still some work to do on them).
> Even when I try `load_nifti` after a simple call to dcm2niix on my `b1_maps` folder as follows:
> 
> ```
> loaded_nifti = load_nifti.load_nifti(r'C:\Users\gaspa\Documents\GitHub\shimming-toolbox-py\testing_data\b1_maps')
> ```
> 
> It returns a 5d array that is a 64x2x8x2x64 array instead of (64x64x2x16 in the original nifti).
> 
> ```
> loaded_nifti[0].shape
> (64, 2, 16, 1, 64)
> ```

True enough, the answer should be a 64x64x2x16x1. I'll check it later.

File: shimmingtoolbox/load_nifti.py
Comment (gab-berest) [reply shimmingtoolbox/load_nifti.py]: I changed the function so it will detect if there are multiple runs in the same folder and ask for which one to load. It should be clarified for later on.

**Comment by gab-berest** [comment]
I added a case that checks the run number and asks for the one to analyze and an argument that tells if phase or magnitude needs to be loaded (checked from ImageComments like dicom_to_nifti.

**Review by jidicula — CHANGES_REQUESTED** [review]
Could you also add automodule stubs for `load_nifti` in the API documentation? It'll be a one-liner in `docs/source/6_api_reference/api.rst`, and you can follow the format of existing stubs for modules like `download` or `utils`.

**Comment by gab-berest** [comment]
> 
> 
> Could you also add automodule stubs for `load_nifti` in the API documentation? It'll be a one-liner in `docs/source/6_api_reference/api.rst`, and you can follow the format of existing stubs for modules like `download` or `utils`.

Done

**Comment by gab-berest** [comment]
Is there something to change? @jcohenadad @po09i @jidicula

File: shimmingtoolbox/load_nifti.py
Comment (jidicula) [inline shimmingtoolbox/load_nifti.py]: Other than this inner for-loop, the enclosing for loop is a duplicate of line 114-120. This is out of scope for the current PR, but is a prime candidate for refactoring in the future. The duplicated code could be factored out into a function taking a boolean param that only runs the inner for-loop for the appropriate case. Also, depending on the size of `range(len(run_list[select_run]))` and `range(nifti_init[0][0].shape[3])`, the `map` function could be used to achieve better performance.

**Review by jidicula — APPROVED** [review]
Doc stubs look ok to me!

File: shimmingtoolbox/load_nifti.py
Comment (jcohenadad) [inline shimmingtoolbox/load_nifti.py:27]: "folder containing acquisitions" --> what do you mean by "acquisitions"? do you mean DICOM files?

File: shimmingtoolbox/load_nifti.py
Comment (jcohenadad) [inline shimmingtoolbox/load_nifti.py:19]:
```suggestion
        path_data (str): Path to the directory containing the file(s) to load
```

File: shimmingtoolbox/load_nifti.py
Comment (jcohenadad) [inline shimmingtoolbox/load_nifti.py:17]: i find it confusing: here it says the function loads from a nifti file, but `path_data` is expecting a folder (not a file)

File: shimmingtoolbox/load_nifti.py
Comment (jcohenadad) [inline shimmingtoolbox/load_nifti.py:22]:
```suggestion
        nibabel.Nifti1Image.Header: List containing headers for every Nifti file
```

File: shimmingtoolbox/load_nifti.py
Comment (po09i) [inline shimmingtoolbox/load_nifti.py:161]: This seems like an edge case. "ImageComments" won't always have in the description if it is phase data or not. For example, our testing data does not even have the field ImageComments and uses the more useful ImageType to differentiate between phase, mag and complex.

File: shimmingtoolbox/load_nifti.py
Comment (po09i) [inline shimmingtoolbox/load_nifti.py:168]: This will obviously need to be removed before merging

File: shimmingtoolbox/load_nifti.py
Comment (gaspardcereza) [reply shimmingtoolbox/load_nifti.py:161]: It would also maybe be better to check for "*phase*" instead of just "phase" in case the user decides to write some other comments in the "ImageComments" tag.

File: shimmingtoolbox/load_nifti.py
Comment (gab-berest) [reply shimmingtoolbox/load_nifti.py:164]: Yes it was while I was testing

File: shimmingtoolbox/load_nifti.py
Comment (gab-berest) [reply shimmingtoolbox/load_nifti.py:157]: I was checking the ImageType before, but this field is also not always present in the json file outputted by dcm2bids. I read the other issues to find that @gaspardcereza was adding the phase, mag, complex tag in the ImageComments. That's why I used it.

File: shimmingtoolbox/load_nifti.py
Comment (gab-berest) [reply shimmingtoolbox/load_nifti.py:27]: You can specify a directory wich contains multiple other directories with the data files. It just tells you that you need to choose one of the folders contained in the given path. Should I rename it? It was named like this in the MATLAB version.

File: shimmingtoolbox/load_nifti.py
Comment (gab-berest) [reply shimmingtoolbox/load_nifti.py:17]: Yes it should say: "from a directory containing NIFTI type files"

File: shimmingtoolbox/load_nifti.py
Comment (gab-berest) [reply shimmingtoolbox/load_nifti.py]: It's true that I doubled my code. I will try to remove it. I don't think a separate function will be usefull since it will be used only these two times and that we can merge the duplicated code into a more condensed code.

**Comment by gab-berest** [comment]
@po09i @jcohenadad Are there any other questions to merge since a subsequent PR might change some feature for PyBIDS?

**Comment by jcohenadad** [comment]
> @po09i @jcohenadad Are there any other questions to merge since a subsequent PR might change some feature for PyBIDS?

all good for me

**Review by gaspardcereza — APPROVED** [review]

**Review by po09i — APPROVED** [review]

**Pull request merged**
